tests/generative/test_query_model.py

```diff
@@ -65,6 +65,7 @@
 settings = dict(
     max_examples=(int(os.environ.get("GENTEST_EXAMPLES", 100))),
     deadline=None,
+    suppress_health_check=[hyp.HealthCheck.filter_too_much],
```

Comment:
Do you understand why this is needed? It's just the filtering to ensure that we actually hit the database that we're doing too much of?

Reply:
I think it must be - that's the only filtering we have.  I'd guessed that it was just because there are more possible operations now so it generates more that don't hit the database, but I may have misunderstood.  It's still managing to pass the comprehensive test with only about 300 examples

Reply:
> It's still managing to pass the comprehensive test with only about 300 examples

Oh, that is cool. What an improvement!

Reply:
> I'd guessed that it was just because there are more possible operations now so it generates more that don't hit the database

Yes, it kind of must be that, mustn't it, but it seem surprising (to me) because you haven't introduced any new non-database terminal nodes. Have you looked at sample of the output queries? I think it's worth doing that from time to time just to make sure you're not missing something. For example if we weren't able to generate `date`-typed `SelectColumn` nodes for some reason, then all the `date`-typed terminals would be `Value` nodes -- and that might result in more filtering.

Reply:
I wonder if what's happening here is that we're over-using `one_row_per_patient_frame_or(frame))`. Each time we do that we give Hypothesis the chance to "drop down" from an event frame to patient frame; and the more operations we stack up that use that function, the greater the chance that it will eventually choose to do so. Once it's got a patient frame it's highly incentivised to choose a `Value` (the simplest of all possible operations).

I think it might be worth only using `one_row_per_patient_from_or` for binary operations where it's somewhat interesting to combine event and patient frames in a single operation. For unary operations I think it's pointless: if we'd wanted a patient frame we could have just chosen a patient frame in the first place.

Reply:
Expanding this thought further: once Hypothesis has chosen the domain/frame for a series I don't think we should be giving it the option to change its mind. This means that for unary operations we should just pass the frame straight through. For binary operations, we can always use a patient frame _in addition to_, but not instead of, the original frame without changing the domain. So we want something like:
```py
@st.composite
def binary_operation_with_types(draw, lhs_type, rhs_type, frame, operator_func):
    other_frame = draw(one_row_per_patient_frame_or(frame))
    # I don't know off the top of my head how to do this in Hypothesis, but I
    # assume it's possible
    lhs_frame, rhs_frame = made_up_funcs.shuffle([frame, other_frame])
    lhs = draw(series(lhs_type, lhs_frame))
    rhs = draw(series(rhs_type, rhs_frame))
    return operator_func(lhs, rhs)
```

Reply:
I think this makes sense to me (the not using `one_row_per_patient_frame_or` everywhere, at least).  

> For binary operations, we can always use a patient frame in addition to, but not instead of, the original frame without changing the domain

Can you explain this a bit more?  I'm not sure I fully understand what domains are.

Reply:
I changed all `one_row_per_patient_frame_or` in unary operators to just pass on the original frame, and changed the function above like this (just using random to shuffle the two frames):
```
    @st.composite
    def binary_operation_with_types(draw, lhs_type, rhs_type, frame, operator_func):
        other_frame = draw(one_row_per_patient_frame_or(frame))
        frames = [frame, other_frame]
        random.shuffle(frames)
        lhs_frame, rhs_frame = frames
        lhs = draw(series(lhs_type, lhs_frame))
        rhs = draw(series(rhs_type, rhs_frame))
        return operator_func(lhs, rhs)
```
And it still complains that it's filtering out too much with 100 or 200 examples, but with 300 it's OK. 
**However** I've realised that the same is true before I made these changes - i.e. with `one_row_per_patient_frame_or` used everywhere, it complains about the filtering with 200 examples, but not with 300.

 If the binary operator also just uses the frame it's passed for both lhs and rhs frames, it doesn't complain with even 100 examples (which is to be expected, I guess).

Reply:
Yeah, this is pretty much the hardest bit to understand of the query model — and the bit which took us the longest to work out.

You can think of a series' domain as the answer to the question "which other series can I compose this with?". For a many-rows-per-patient series, its domain is the frame its values were drawn from. For all one-row-per-patient series (which include static values) it's a sort of shared base, or root, domain we call the "patient domain".

The basic rule is that you can't combine different many-rows-per-patient domains, but you _can_ combine a many-rows-per-patient domain with a one-row-per-patient domain.
 This is what's behind the `one_row_per_patient_or(frame)` strategy: it allows us to test mixing a many-rows-per-patient frame with a one-row-per-patient one.

But if we call it twice we might end up throwing away the original frame, and I think that's part of the problem here. So my suggestion was to always keep the original frame and then decide if we're going to combine it with a series drawn from the same frame or whether to mix it with a one-row-per-patient one.

Does that make any sense?

Reply:
I was conscious when I wrote this that the implementation would "demote" the domain to the patient domain some proportion of the time. I didn't like it all that much, but since it was valid I left it be.

But you're right, Dave, that

1) it's not really consistent with contract of the concrete strategies (because the demotion is observable from outside, rather than just being an implementation detail); but it would be (consistent and just an implementation detail) if only _one_ of the inputs to a binary operator was chosen to be demoted
2) it could be causing this problem

So I agree that the right thing to do is drop the demotion for unary operators and never demote both of the inputs for binary operators.

Reply:
I'm not sure how this is going to behave with `random.shuffle` as it's not under Hypothesis' control and is non-deterministic. I think you might be able to do it with:
```py
swap = draw(st.booleans())
lhs_frame, rhs_frame = (frame, other frame) if not swap else (other_frame, frame)
```

Reply:
I guess that Becky's latest findings support the general thrust here:

> If the binary operator also just uses the frame it's passed for both lhs and rhs frames, it doesn't complain with even 100 examples

But it's a bit sad that the approach of only demoting one and only sometimes doesn't fix the problem.

I suppose we could try biasing the distribution further (trivially but clunkily by putting more than one instance of `frame` into the shuffled list).

Reply:
A couple of points on the new implementation as it stands.

Can we inline `one_row_per_patient_frame_or()`, if that's the only place it's being used now? It will make it much clearer what's going on, I think. I think that will also allow a clearer implementation where we chose between three options for the inputs: `frame, patient_frame`, `patient_frame, frame` or `frame, frame`.

You shouldn't use `random` because it clashes with Hypothesis' reproducibility needs. But Hypothesis has a replacement for it that you can use if you need. Here I suspect that you don't and you can compose what you need out of existing strategies instead, like the use of `one_of()` in `one_row_per_patient_frame_or()`.

Reply:
Ooh, further thought here: the root problem is that Hypothesis is too keen to pick `Values` (because they're simple) but by doing so it ends up building expressions which don't touch the database at all.

What if we said that the _only_ place it should generate `Values` is in combination with series which aren't `Values`? I don't think that's overlay restrictive: operations that only involve static values are pointless; we can resolve them statically before compiling to SQL if we want to.

Changing this would involve taking `value` out of the `series_constraints` dict and then amending the logic in `binary_operation_with_types` to something that chooses whether we're combining the original frame with either itself, or a new one-row-per-patient frame, or a value; and then draws series accordingly.

I think if we did that we might even be able to do away with the "uses database" check because we'd never generate examples which failed it.

Reply:
Yes, you're right about `Case` and `In`. I think any operation involving more than one argument needs some generalisation of the logic for binary operations.

Just experimenting with what Hypothesis lets us do, I think something like this might work (though it's not particularly elegant):
```py
@st.composite
def binary_operation_with_types(draw, lhs_type, rhs_type, frame, operator_func):
    other_frame = draw(st.one_of(one_row_per_patient_frame(), st.just(frame)))
    if is_one_row_per_patient_frame(other_frame):
        other_strategy = draw(st.sampled_from([value, series]))
    else:
        other_strategy = series
    (lhs_frame, lhs_strategy), (rhs_frame, rhs_strategy) = st.permutations(
        [(frame, series), (other_frame, other_strategy)]
    )
    lhs = draw(lhs_strategy(lhs_type, lhs_frame))
    rhs = draw(rhs_strategy(rhs_type, rhs_frame))
    return operator_func(lhs, rhs)
```

Reply:
I've updated this PR now with a version of @evansd's suggestion above (had to use an alternative to `st.permutations` because you can't unpack a a strategy), and it seems to be doing the right thing - no need for the filter check suppression, and it's still passing the comprehensive tests with 300 examples.

From the last run (300 examples):
```
247 unique input combinations

173 unique queries

with a total of 1564 nodes

with this node count distribution
  2	18
  3	7
  4	20
  5	22
  6	19
  7	26
  8	5
  9	10
 10	4
 11	5
 12	5
 13	2
 14	3
 15	5
 16	1
 18	5
 19	3
 20	1
 22	2
 23	1
 25	1
 26	1
 27	1
 30	2
 44	1
 45	2
 71	1

and these node types
Value                    387
SelectPatientTable       257
SelectColumn             246
Multiply                 68
Subtract                 65
Add                      57
ToFirstOfMonth           56
DateAddMonths            46
Negate                   46
Or                       45
DateAddYears             35
LE                       34
DateAddDays              31
ToFirstOfYear            31
DateDifferenceInMonths   27
MonthFromDate            16
DateDifferenceInDays     15
DateDifferenceInYears    15
YearFromDate             14
Count                    13
GT                       10
LT                       9
And                      7
DayFromDate              6
SelectTable              5
GE                       4
IsNull                   4
Not                      4
EQ                       3
Exists                   3
NE                       3
Filter                   2

166 unique datasets

with this size distribution
  0	1
  1	41
  2	7
  3	5
  4	7
  5	5
  6	6
  7	5
  8	2
  9	9
 10	12
 11	3
 12	5
 13	8
 14	13
 15	13
 16	3
 17	5
 19	2
 20	2
 21	5
 22	6
 26	1

7 errors ignored
```

Reply:
I don't think the logic is quite right (or it is right but it's unclear so I'm misunderstanding it).

I think that the logic should be that of the two inputs to a binary operator:
* one input should always be an arbitrary series on the specified frame
* the other input should be _either_ an arbitrary series on the specified frame _or_ an arbitrary series on a patient-domain frame _or_ a `Value`

(With the caveat that if the frame is already patient-level then that second case collapses to just two possibilities.)

And we remove `Value` from the list of arbitrary series.

Reply:
> I think that we may need to allow for values Case as well, and In in case that is given special treatment.

I thought this meant that we *shouldn't* remove `value` from the `series_constraints`.  Did you mean that Case/In (when we get to them) need to be treated differently to other binary operators?

Reply:
> it seems to be doing the right thing

Well that's encouraging.


> had to use an alternative to `st.permutations` because you can't unpack a strategy

Ah yeah, I find it so hard to keep it straight in my head whether I'm dealing with a strategy or the result of a strategy. I think you might be able to do with with `draw(st.permutations(...))` but what you've got does the trick as well.


> I don't think the logic is quite right (or it is right but it's unclear so I'm misunderstanding it).

I'm 80% confident it's right, and 100% confident it's unclear :smile: 

I think it's equivalent to what you say, but instead of picking directly from the trio `(original-frame, patient-frame, value)` it first picks from `(original-frame, patient-frame)` and then splits the second branch into `(patient-frame, value)`.

I've just experimented with some alternatives but it's quite tricky to express this perspicaciously in Hypothesis.

I don't think this comment's quite right though:
```
        # we can combine this frame with a series drawn from:
        # - a new one-row-per-patient-frame
        # - this frame itself
        # If it's already a one-row-per-patient-frame, it uses the database, so can
        # be combined with a value or a series
        # If it's not a one-row-per-patient-frame, avoid combining with a value so we
        # don't end up with operations that don't hit the database
```

Reply:
> I think that the logic should be that of the two inputs to a binary operator:
> 
>  - one input should always be an arbitrary series on the specified frame
>  - the other input should be either an arbitrary series on the specified frame or an arbitrary series on a patient-domain frame or a Value

I think this is what's happening.  I'll see if I can make it clearer.

Reply:
> Did you mean that Case/In (when we get to them) need to be treated differently to other binary operators?

Well I don't think of `Case` as a binary operator. I guess it does take two arguments, but one of those is a dictionary with arbitrary many series in it. So we're going to have to generalise the logic for binary operators somehow. I think this means that we'll need to pick at least one "condition" series to use the original frame, and then the rest can use the original frame, or a one-row-per-patient frame, or a value.

`In` is a bit different because we don't currently have a way of specifying anything _other_ than a `Value` on the RHS.

But I propose worrying about both of these when we come to them. Either way this means removing `value` from `series_constraints`.

Reply:
On second thoughts, I do think the logic is what we want.  I've tried to improve the comment.

Reply:
> I assume you mean that if it's already patient-level then the only 2 options are a series on the specified frame or a value. Why could it not also be a series on another patient-level frame?

@rebkwok You're quite right about this.